File: solved/p187/p187.py
#!/usr/bin/python

from primes import get_primes

# pretty simple, just determine for each prime, p, the set of primes
# {q: q >= p && p*q < Nmax}.  Each of these items contributes a single
# count to the result.

NMAX = 100000000

primes = get_primes(NMAX)

result = 0
lastptr = -1
for (ip, p) in enumerate(primes):
    q = primes[lastptr]
    while p*q >= NMAX:
        lastptr -= 1
        q = primes[lastptr]

    if q < p:
        break

    num_items = len(primes) + 1 + lastptr - ip
    result += num_items

print(result)

File: solved/p187/p187.c
#include <prime.h>
#include <stdio.h>

#define NMAX 100000000

int main(void) {
    long np, *p;
    primes(NMAX, &p, &np);

    long lastptr = np-1;

    int i;
    long result = 0;

    for(i = 0; i < np; i++) {
        long qmax = NMAX / p[i];

        while(p[lastptr] > qmax) {
            lastptr--;
        }

        if(lastptr < i)
            break;

        result += lastptr - i + 1;
    }

    printf("result: %ld\n", result);
    return 0;
}
